tests/flow/test_optimizations_plan.py

```diff
@@ -263,4 +263,13 @@ def test14_distinct_aggregations(self):
         # This query should emit the same result.
         self.env.assertEqual(resultset, expected)
 
-
+    def test15_test_splitting_cartesian_product(self):
+        query = """MATCH (p1), (p2), (p3) WHERE p1.name <> p2.name AND p2.name <> p3.name RETURN DISTINCT p2.name ORDER BY p2.name"""
+        executionPlan = graph.execution_plan(query)
+        self.env.assertEqual(2, executionPlan.count("Cartesian Product"))
+        expected = [['Ailon'],
+                    ['Alon'],
+                    ['Boaz'],
+                    ['Roi']]
+        resultset = graph.query(query).result_set
+        self.env.assertEqual(resultset, expected)
```

Comment:
Let's add another test that triggers the `(lhs_resolving_stream == rhs_resolving_stream)` condition, such as:
`MATCH (p1), (p2), (p3) WHERE p1.name <> p2.name AND ID(p1) <> ID(p2) RETURN DISTINCT p2.name ORDER BY p2.name`